src/utils/stats_alloc/src/task_local.rs

```diff
@@ -0,0 +1,203 @@
+                // happens before decreasing the counter, which happens before this fence, which
+                // happens before the deletion of the counter.
+                fence(Ordering::Acquire);
+                unsafe { Box::from_raw_in(bytes.as_mut_ptr(), System) };
+            }
+        }
+    }
+
+    #[inline(always)]
+    pub fn val(&self) -> usize {
+        self.0
+            .as_ref()
+            .expect("bytes is invalid")
+            .load(Ordering::Relaxed)
+    }
+}
+
+task_local! {
+    static BYTES_ALLOCATED: TaskLocalBytesAllocated;
+}
+
+pub async fn allocation_stat<Fut, T, F>(future: Fut, interval: Duration, mut report: F) -> T
+where
+    Fut: Future<Output = T>,
+    F: FnMut(usize),
+{
+    BYTES_ALLOCATED
+        .scope(TaskLocalBytesAllocated::new(), async move {
+            // The guard has the same lifetime as the counter so that the counter will keep positive
+            // in the whole scope. When the scope exits, the guard is released, so the counter can
+            // reach zero eventually and then `drop` itself.
+            let _guard = Box::new(114514);
+            let monitor = async move {
+                let mut interval = tokio::time::interval(interval);
+                loop {
+                    interval.tick().await;
+                    BYTES_ALLOCATED.with(|bytes| report(bytes.val()));
+                }
+            };
+            let output = tokio::select! {
+                biased;
+                _ = monitor => unreachable!(),
+                output = future => output,
+            };
+            output
+        })
+        .await
+}
+
+#[inline(always)]
+fn wrap_layout(layout: Layout) -> (Layout, usize) {
+    debug_assert_ne!(layout.size(), 0, "the size of layout must be non-zero");
+
+    let (wrapped_layout, offset) = Layout::new::<TaskLocalBytesAllocated>()
+        .extend(layout)
+        .expect("wrapping layout overflow");
+    let wrapped_layout = wrapped_layout.pad_to_align();
+
+    (wrapped_layout, offset)
+}
+
+pub struct TaskLocalAlloc<A>(pub A);
+
+unsafe impl<A> GlobalAlloc for TaskLocalAlloc<A>
+where
+    A: GlobalAlloc,
+{
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let (wrapped_layout, offset) = wrap_layout(layout);
+
+        BYTES_ALLOCATED
+            .try_with(|&bytes| {
+                bytes.add_unchecked(layout.size());
+                let ptr = self.0.alloc(wrapped_layout);
+                *ptr.cast() = bytes;
+                ptr.wrapping_add(offset)
+            })
+            .unwrap_or_else(|_| {
+                let ptr = self.0.alloc(wrapped_layout);
+                *ptr.cast() = TaskLocalBytesAllocated::invalid();
+                ptr.wrapping_add(offset)
+            })
+    }
+
+    unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
+        let (wrapped_layout, offset) = wrap_layout(layout);
+        let ptr = ptr.wrapping_sub(offset);
+
+        let bytes: TaskLocalBytesAllocated = *ptr.cast();
```

Comment:
What's the purpose of this? Those bytes are going to be dealloc anyway. In the first place, why do we prefix every alloc with this type? There doesn't seem to be a point, all alloced bytes can simply be counted by the task local var. 

Of course, we can encounter the scenario where another task frees the allocated memory etc. But maybe this is out of scope.

Reply:
Anw, I think we can get rid of this extra allocation and rely only on task local var. Anw, we should see if unit test passes on current and if not check if this change will pass.

Reply:
> There doesn't seem to be a point, all alloced bytes can simply be counted by the task local var.

If we allocated some bytes in actor 1, send them to actor 2 through the channel, then actor 2 deallocates them, the substraction will be applied on the actor 2's task-local and the result becomes incorrect.